============
CAMP Binning
============


.. image:: https://readthedocs.org/projects/camp-binning/badge/?version=latest
        :target: https://camp-binning.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/badge/version-0.1.0-brightgreen


Overview
--------

This module is designed to function as both a standalone MAG binning pipeline as well as a component of the larger CAP2/CAMP metagenome analysis pipeline. As such, it is both self-contained (ex. instructions included for the setup of a versioned environment, etc.), and seamlessly compatible with other CAMP modules (ex. ingests and spawns standardized input/output config files, etc.). 

As far as the binning procedure goes, the design philosophy is just to replicate the functionality of MetaWRAP (one of the original ensemble `binning methods <https://github.com/bxlab/metaWRAP>`_) with i) better dependency conflict management and ii) improved integration with new binning algorithms. 

Currently, the only binning algorithms wrapped are MetaBAT2 and CONCOCT, but not for long!

Installation
------------

1. Clone repo from `Github <https://github.com/MetaSUB-CAMP/camp_binning>`_. 

2. Set up the conda environment (contains, Snakemake) using ``configs/conda/binning.yaml``. 
    - There are some compatibility issues that I haven't ironed out (bowtie2 due to RedHat's geriatric dependencies), so you may have to substitute in your own version. 

3. Make sure the installed pipeline works correctly. ``pytest`` only generates temporary outputs so no files should be created.
::
    cd camp_binning
    conda env create -f configs/conda/binning.yaml
    conda activate camp_binning
    pytest .tests/unit/

Using the Module
----------------

**Input**: ``/path/to/samples.csv`` provided by the user.

**Output**: 1) An output config file summarizing the locations of 2) the MAGs generated by MetaBAT2, CONCOCT, and VAMB. 

- ``/path/to/work/dir/binning/final_reports/samples.csv`` for ingestion by the next module (ex. quality-checking)
- ``/path/to/work/dir/binning/*/sample_name/``, where ``*`` is either ``1_metabat2`` or ``2_concoct``, the directories containing FastAs (``*.fa``) of MAGs inferred by MetaBAT2 and CONCOCT respectively

**Structure**:
::
    └── workflow
        ├── Snakefile
        ├── binning.py
        ├── utils.py
        └── __init__.py
- ``workflow/binning.py``: Click-based CLI that wraps the ``snakemake`` and unit test generation commands for clean management of parameters, resources, and environment variables.
- ``workflow/Snakefile``: The ``snakemake`` pipeline. 
- ``workflow/utils.py``: Sample ingestion and work directory setup functions, and other utility functions used in the pipeline and the CLI.

1. Make your own ``samples.csv`` based on the template in ``configs/samples.csv``. Sample test data can be found in ``.tests/unit/map_sort/data/work_dir/tmp/``.
    - ``ingest_samples`` in ``workflow/utils.py`` expects Illumina reads in FastQ (may be gzipped) form and de novo assembled contigs in FastA form
    - ``samples.csv`` requires either absolute paths or symlinks relative to the directory that the module is being run in

2. Update the relevant ``metabat2``, ``concoct``, and ``vamb`` parameters in ``configs/parameters.yaml``.

3. Update the computational resources available to the pipeline in ``resources.yaml``. 

4. To run CAMP on the command line, use the following, where ``/path/to/work/dir`` is replaced with the absolute path of your chosen working directory, and ``/path/to/samples.csv`` is replaced with your copy of ``samples.csv``. 
    - The default number of cores available to Snakemake is 1 which is enough for test data, but should probably be adjusted to 10+ for a real dataset.
    - Relative or absolute paths to the Snakefile and/or the working directory (if you're running elsewhere) are accepted!
::
    python /path/to/camp_binning/workflow/binning.py \
        (-c max_number_of_local_cpu_cores) \
        -d /path/to/work/dir \
        -s /path/to/samples.csv
* Note: This setup allows the main Snakefile to live outside of the work directory.

5. To run CAMP on a 5. To run CAMP on a job submission cluster (for now, only Slurm is supported), use the following.
    - ``--slurm`` is an optional flag that submits all rules in the Snakemake pipeline as ``sbatch`` jobs. 
    - In Slurm mode, the ``-c`` flag refers to the maximum number of ``sbatch`` jobs submitted in parallel, **not** the pool of cores available to run the jobs. Each job will request the number of cores specified by threads in ``configs/resources/slurm.yaml``.
::

    sbatch -J jobname -o jobname.log << "EOF"
    #!/bin/bash
    python /path/to/camp_binning/workflow/binning.py --slurm \
        (-c max_number_of_parallel_jobs_submitted) \
        -d /path/to/work/dir \
        -s /path/to/samples.csv
    EOF

6. After checking over ``final_reports/`` and making sure you have everything you need, you can delete all intermediate files to save space. 
::

    python /path/to/camp_binning/workflow/binning.py cleanup \
        -d /path/to/work/dir \
        -s /path/to/samples.csv

7. If for some reason the module keeps failing, CAMP can print a script containing all of the remaining commands that can be run manually. 
::

    python /path/to/camp_binning/workflow/binning.py --dry_run \
        -d /path/to/work/dir \
        -s /path/to/samples.csv > cmds.txt
    python /path/to/camp_binning/workflow/binning.py commands cmds.txt

Extending the Module
--------------------

We love to see it! This module was partially envisioned as a dependable, prepackaged sandbox for developers to test their shiny new tools in. 

These instructions are meant for developers who have made a tool and want to integrate or demo its functionality as part of the standard binning workflow, or developers who want to integrate an existing tool. 

1. Write a module rule that wraps your tool and integrates its input and output into the pipeline. 
    - This is a great `Snakemake tutorial <https://bluegenes.github.io/hpc-snakemake-tips/>`_ for writing basic Snakemake rules.
    - If you're adding new tools from an existing YAML, use ``conda env update --file configs/conda/existing.yaml --prune``.
    - If you're using external scripts and resource files that i) cannot easily be integrated into either `utils.py` or `parameters.yaml`, and ii) are not as large as databases that would justify an externally stored download, add them to ``workflow/ext/`` or ``workflow/ext/scripts/`` and use ``rule external_rule`` as a template to wrap them. 
2. Update the ``make_config`` in ``workflow/Snakefile`` rule to check for your tool's output files. Update ``samples.csv`` to document its output if downstream modules/tools are meant to ingest it. 
    - If you plan to integrate multiple tools into the module that serve the same purpose but with different input or output requirements (ex. for alignment, Minimap2 for Nanopore reads vs. Bowtie2 for Illumina reads), you can toggle between these different 'streams' by setting the final files expected by ``make_config`` using the example function ``workflow_mode``.
    - Update the description of the ``samples.csv`` input fields in the CLI script ``workflow/binning.py``. 
3. If applicable, update the default conda config using ``conda env export > config/conda/binning.yaml`` with your tool and its dependencies. 
    - If there are dependency conflicts, make a new conda YAML under ``configs/conda`` and specify its usage in specific rules using the ``conda`` option (see ``first_rule`` for an example).
4. Add your tool's installation and running instructions to the module documentation and (if applicable) add the repo to your `Read the Docs account <https://readthedocs.org/>`_ + turn on the Read the Docs service hook.
5. Run the pipeline once through to make sure everything works using the test data in ``test_data/`` if appropriate, or your own appropriately-sized test data. Then, generate unit tests to ensure that others can sanity-check their installations.
    * Note: Python functions imported from ``utils.py`` into ``Snakefile`` should be debugged on the command-line first before being added to a rule because Snakemake doesn't port standard output/error well when using ``run:``.
::
    python /path/to/camp_binning/workflow/binning.py (--unit_test) \
        -d /path/to/work/dir \
        -s /path/to/samples.csv

6. Increment the version number of the modular pipeline.
::
    bump2version --allow-dirty --commit --tag major workflow/__init__.py \
                 --current-version A.C.E --new-version B.D.F

7. If you want your tool integrated into the main CAP2/CAMP pipeline, send a pull request and we'll have a look at it ASAP! 
    - Please make it clear what your tool intends to do by including a summary in the commit/pull request (ex. "Release X.Y.Z: Integration of tool A, which does B to C and outputs D").

Bugs
----

There is a dependency error that hasn't been addressed yet, namely ``bowtie2`` in the main ``camp_binning`` conda environment, which has conflicting C++ and Perl dependencies with some other packages.

Credits
-------

Credits
-------

* This package was created with `Cookiecutter <https://github.com/cookiecutter/cookiecutter>`_ as a simplified version of the `project template <https://github.com/audreyr/cookiecutter-pypackage>`_.
* This module is heavily inspired by four Snakefiles from `MAG Snakemake workflow <https://github.com/Finn-Lab/MAG_Snakemake_wf>`_ (Saheb Kashaf et al. 2021).
* The MAG N50, size, and GC calculation rule was adapted from a script in `MetaWRAP <https://github.com/bxlab/metaWRAP>`_. 
* Free software: MIT 
* Documentation: https://binning.readthedocs.io. 

